cf-agent/cf-agent.c

```diff
@@ -893,13 +894,57 @@ static void FreeFixedStringArray(int size, char **array)
 
 /*******************************************************************/
 
+/**
+ * Gracefully terminate all children processes.
+ */
+static void CleanupProcessTree(void)
+{
+#ifndef _WIN32
+    ClearProcessTable(); /* Make sure old table is not cached */
+
+    Item *procs = NULL;
+    pid_t this, parent = getppid();
+    if (LoadProcessTable())
+    {
+        ProcessSelect ps = PROCESS_SELECT_INIT;
+        ps.max_pgid = ps.min_pgid = getpgrp();
+        this = ps.max_ppid = ps.min_ppid = getpid();
+        procs = SelectProcesses(".*", &ps, true);
+    }
+
+    for (Item *p = procs; p != NULL; p = p->next)
+    {
+        pid_t child = p->counter;
+        if (child == parent || child == this)
+        {
+            continue;
+        }
+
+        if (GracefulTerminate(child, p->time))
+        {
+            Log(LOG_LEVEL_VERBOSE,
+                "Successfully terminated child process %d", child);
+        }
+        else
+        {
+            // This is not an error.
+            Log(LOG_LEVEL_VERBOSE,
+                "Failed to terminate child process '%d'", child);
+        }
+    }
+
+    DeleteItemList(procs);
+#endif
+}
+
 static void ThisAgentInit(void)
 {
     char filename[CF_BUFSIZE];
 
 #ifdef HAVE_SETSID
     setsid();
 #endif
+    RegisterCleanupFunction(CleanupProcessTree);
```

Comment:
Could we combine this with the cleanup function that terminates the processes of custom promise modules?